core/templates/dev/head/services/StateTopAnswersStatsService.ts

```diff
@@ -132,29 +127,29 @@ oppia.factory('StateTopAnswersStatsService', [
         return isInitialized;
       },
 
+      /** @returns {string[]} - list of state names with recorded stats. */
+      getStateNamesWithStats: function() {
+        return Object.keys(workingStateTopAnswersStats);
+      },
+
       /**
        * @returns {boolean} - Whether the cache contains any answers for the
        * given state.
        */
       hasStateStats: function(stateName) {
-        return isInitialized &&
-          stateTopAnswersStatsCache.hasOwnProperty(stateName);
-      },
-
-      /** @returns {string[]} - list of state names with recorded stats. */
-      getStateNamesWithStats: function() {
-        return Object.keys(stateTopAnswersStatsCache);
+        return this.isInitialized() &&
+          this.getStateNamesWithStats().includes(stateName);
```

Comment:
Wouldn't workingStateTopAnswersStats.hasOwnProperty(stateName) be faster? (O(1) lookup rather than O(N))

Exps can have lots of states, so if this is called frequently the speedup may be meaningful. Probably helps prevent needless jankiness.

Reply:
Yup good point, done.